### PR TITLE
[ADD] core: ORM-level internal grouping utility

### PR DESCRIPTION
**Pull request by xmo-odoo**
When looking to fetch grouped data from the database, `read_group` should work(-ish), especially with support for `array_agg`. This also means this need should be more or less solved around RPC: client can either `read_group` or group on the client side.

This leaves a glaring hole in the fabric: there's no convenient tools to apply grouping to recordsets. `itertools.groupby` exists, but it requires that the input be grouped by the key function, and it yields an iterable of items which is not the most convenient for recordsets.

This here utility:

- is exclusive to recordsets
- returns a mapping of grouping keys to subsets of the input recordset
- keeps the prefetching of the source recordset
- allows grouping on a field, or an arbitrary (callable) key
- should work even on "new records" (aka should be suitable for onchanges / arbitrary compute functions)

The method was made non-RPC as it seems unnecessary (and even an anti-pattern) in that context: although there is something to be said for its ability to group on non-stored fields, it seems like a much better idea for the client to `read` all the data they need, then group that client-side on whatever criteria they are interested in.

A few possible points of contention:

- remove `_` prefix, being callable over RPC is useless but technically doesn't hurt.
- method name, `partition` seemed nice but it's usually binary (true/false), and we already have one in tools, `group_by` seems a bit too overloaded or something we might want to use for a replacement of read_group (also conflicts with itertools). "group on" is common haskell lingo, apparently.
- ability to work with chained fields, for convenience compatibility with m2o, seems unnecessary.

### CODE REVIEW

**Review by ryv-odoo — APPROVED** [review]

File: odoo/models.py [hunk 1]
```diff
@@ -5851,6 +5851,33 @@ def filtered(self, func):
...
+    def _group_on(self, key):
+        """Eagerly groups the records of ``self`` by the ``key``, returning a
+        dict from the ``key``'s result to recordsets. All the resulting
+        recordsets are guaranteed to be part of the same prefetch-set.
+
+        Provides a convenience method to partition existing recordsets without
+        the overhead of a :meth:`~.read_group`, but performs no aggregation.
+
+        .. note:: unlike :func:`itertools.groupby`, does not care about input
+                  ordering, however the tradeoff is that it can not be lazy
```
Comment (hoangtrann) [inline odoo/models.py:5863]: valuable note 💯 

most of the time, I see the mistake of using `itertools.groupby` without sorting first, causing unexpected result

File: odoo/models.py [hunk 2]
```diff
@@ -5851,6 +5851,33 @@ def filtered(self, func):
...
+
+        :param key: either a callable from a :class:`Model` to a (hashable)
+                    value, or a field name. In the latter case, it is equivalent
+                    to ``itemgetter(key)`` (aka the named field's value)
+        :type key: callable | str
+        :rtype: dict
+        """
+        if isinstance(key, str):
+            key = itemgetter(key)
```
Comment (hoangtrann) [inline odoo/models.py:5872]: Suggest to support grouping by multiple keys, either by provided a string of fields, separated by comma, or to input `key` as `list` or `tuple`

```suggestion
            key = itemgetter(*(k.strip() for k in key.split(",")))
```

**Comment by rousseldenis** [comment]
Sound great !

**Review by ryv-odoo — APPROVED** [review]
Hello @xmo-odoo ,
A last detail: can you change the name to be `grouped` as we discussed with @rco-odoo (same as `filtered`, `mapped`) ?

**Comment by xmo-odoo** [comment]
@ryv-odoo thanks for the reminder, I'd completely forgotten.

**Comment by ryv-odoo** [comment]
LGTM
@robodoo r+ rebase-ff

**Pull request closed without merging**